flink-runtime/src/main/java/org/apache/flink/runtime/state/AbstractChannelStateHandle.java

```diff
@@ -46,6 +46,11 @@
 	private final List<Long> offsets;
 	private final long size;
 
+	/**
+	 * The original subtask index before rescaling recovery.
+	 */
+	private int originalSubtaskIndex;
```

Comment:
You probably should serialize this field and add it to the equals/hashCode.

Reply:
It is serialized when transferring the `OperatorSubtaskState` through `Serializable`. It's currently not needed in the checkpoints (as the information can be easily inferred). Is it still necessary from your perspective?
I'd add it to `equals/hashCode` anyways.

Reply:
That means that after recovery it's not valid until explicitly set, right?
I think it's quite error-prone. I'd at least document this behavior and mark it deprecated to give a warning.

Reply:
I made it part of the handle and it's fully serialized now. It's renamed to `subtaskIndex`.